Fix delete a image while saving it, delete successfully but failed to save it

moypray opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Fix Issue:   https://github.com/docker/docker/issues/29653  
Issue Description:
 * 1. Saving more than one images, `docker save -o a.tar aaa bbb`
 * 2. Delete the last image which in saving progress.  `docker rmi bbb`

Espected:
  Saving images operation shouldn't be disturbed. But the real result is that failed
to save image and get an error as below:
  `Error response from daemon: open
/var/lib/docker/image/devicemapper/imagedb/content/sha256/7c24e4d533a76e801662ad1b7e6e06bc1204f80110b5623e96ba2877c51479a1:
**- How I did it**

Analysis:
 1. While saving more than one images, it will get all the image info
from reference/imagestore, and then using the `cached data` to save the
images to a tar file.
 2. But this process doesn't have a resource lock, if a deletion operation comes, the image will be deleted, so saving operation will fail.

Solution:
 Add a `imageStateMgr` to save the image saving status, while before deleting a image, check the image status, if the image is in saving progress, stop the deletion.

**- How to verify it**

First Terminal:
   docker save -o a.tar  aaa  bbb   // especially aaa is a large image.
Another Terminal:
   docker rmi  bbb      // remove operation will remove bbb before save
 
**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
@moypray CI is failing on `gofmt`;

```
14:33:47 These files are not properly gofmt'd:
14:33:47  - image/tarexport/save.go
14:33:47 
14:33:47 Please reformat the above files using "gofmt -s -w" and commit the result.
```

vdemeester reviewed (commented):
/cc @cpuguy83 @tonistiigi @vieux

moypray commented:
It is strange that `TestAPIImagesSaveAndLoad` failed. I have run this TC in my local environment successfully. I will take a look at it further.

moypray commented:
hi @thaJeztah , I have fixed the CI issue: (TestAPIImagesSaveAndLoad) 
I found that when we send `image/<id>/get` API to daemon, docker daemon will stuck on io.Copy();
My partial diff:
```
@@ -226,8 +230,11 @@ func (s *saveSession) save(outStream io.Writer) error {
                return err
        }
        defer fs.Close()
+       logrus.Debugf("WT==>Before io.copy")

        _, err = io.Copy(outStream, fs)
+       logrus.Debugf("WT==>After io.copy")
+
        return err
 }
```

And I got log: (notice that the timestamp, "Before io.copy" is `8:50` and "After io.copy“ is `09:00` I also add a sleep 10 seconds before removing the image)
```
root@localhost:~/workspace/github/docker# cat bundles/17.05.0-dev/test-integration-cli/docker.log | grep "WT"
time="2017-04-06T11:08:50.093440871Z" level=debug msg="WT==> Set ImageID: sha256:cd318112c20662699f09ca7ada1b1a5a5f6f2f8747858b1a64e2421c25654ca2 to true"
time="2017-04-06T11:08:50.093784206Z" level=debug msg="WT==> come to save."
time="2017-04-06T11:08:50.124835579Z" level=debug msg="WT==> save id: sha256:cd318112c20662699f09ca7ada1b1a5a5f6f2f8747858b1a64e2421c25654ca2"
time="2017-04-06T11:08:50.125739408Z" level=debug msg="WT==>Before io.copy"
time="2017-04-06T11:09:00.213792227Z" level=debug msg="WT==>After io.copy"
time="2017-04-06T11:09:00.217955678Z" level=debug msg="WT==> Set ImageID: sha256:cd318112c20662699f09ca7ada1b1a5a5f6f2f8747858b1a64e2421c25654ca2 to false"
time="2017-04-06T11:09:00.217990057Z" level=debug msg="WT=> exprt image done"
root@localhost:~/workspace/github/docker#
```
The `io.Copy` will block the main process util some one read the HTTP Body (when load is executed, it will continue to work), so after `save` operation and try to remove the image, it reports the image is in saving process. 

I just modified the TC, read the Body response out before removing the image, it will be passed.

moypray commented:
The failure on ppc arch is: ( is that a known issue? it is nothing to do with my fix)
```
04:27:43 ----------------------------------------------------------------------
04:27:43 FAIL: check_test.go:95: DockerSuite.TearDownTest
04:27:43 
04:27:43 check_test.go:96:
04:27:43     testEnv.Clean(c, dockerBinary)
04:27:43 environment/clean.go:63:
04:27:43     t.Fatalf("error removing containers %v : %v (%s)", containers, result.Error, result.Combined())
04:27:43 ... Error: error removing containers [e267a1728c8d] : exit status 1 (Error response from daemon: removal of container e267a1728c8d is already in progress
```

tophj-ibm commented:
@moypray yes it's a known issue and not related. I'm rebuilding the job.

aaronlehmann commented on daemon/image_delete.go:
```diff
@@ -401,6 +402,14 @@ func (daemon *Daemon) checkImageDeleteConflict(imgID image.ID, mask conflictType
+			imgID:   imgID,
+			hard:    true,
+			used:    true,
+			message: fmt.Sprintf("image is being saved in progress, try deleting it later"),
```
Why `Sprintf`?

moypray replied:
Yes, I should add the imageID to the message, I will fix it.

aaronlehmann commented on image/state.go:
```diff
@@ -0,0 +1,65 @@
+var (
+	iSR = &imageStateMgr{
+		imgIDMap: make(map[ID]refCount),
+	}
```
Rather than a global variable, the image state should be part of `Daemon`.

Maybe this functionality should be integrated into `image.Store` instead of keeping track of saves separately.

moypray replied:
I think the  image operation just a temporary state. for example when we save/load/rm a image only when the dockerd is working, if dockerd restarted, I think the operation state SHOULD be cleared. It is non-persistent, so I do not save it in daemon or image.Store.

How about others' opinions?

moypray replied:
oh, I misunderstand, we should put it in `image.Store`, and reset it when docker starts up.

OK, Let's have a further look at the solution, then I will fix it.

tonistiigi commented:
Maybe we should just take the layer reference for every image before starting the save operation and release after completed.

cpuguy83 commented:
Also looks like a similar problem to #30696

aaronlehmann commented:
> Maybe we should just take the layer reference for every image before starting the save operation and release after completed.

I like the idea, but I don't think `image_delete.go` has logic to check this right now. Is that right? What would that logic look like, considering that it would have to ignore references by the image it's about to delete?

tonistiigi commented:
@aaronlehmann `image rmi` would delete them, just it wouldn't interrupt `save`. So if some part of the code needs to hold on to a reference of an the image they need to keep instance of `image.Image` and reference to the rootfs.

moypray commented:
> Maybe we should just take the layer reference for every image before starting the save operation and release after completed.

It is a good idea.  If I didn't misunderstand. How about adding a `Lock/Unlock` in image.Store, we could hold the lock before save and rmi operation,

But this solution, `rmi` will become a blocking API, it will wait for `save` operation finished. and maybe it would take a quite long time.

```
type Store interface {
 16         Create(config []byte) (ID, error)
 17         Get(id ID) (*Image, error)
             ......
 28         Lock(ID) error  // lock the image,
 29         Unlock(ID)       // unlock the image..

```

tonistiigi commented:
@moypray I think instead of blocking(or erroring) just delaying the actual delete from disk (https://github.com/docker/docker/pull/32293#issuecomment-292640251) is better ux.

You shouldn't need any new locks. `rmi` will remove the image from the imagestore and that should not interrupt save if it already has acquired a pointer to `image.Image`. The `rmi` part shouldn't require any changes, just in the beginning of the "save" you have to get a list of all images that will be saved and take references to their rootfs.

moypray commented:
@tonistiigi  brilliant

moypray commented:
Sorry for late, I am so busy these days, may I update it later???

mlaventure commented:
ping @moypray in case you got a bit more time now :)

moypray commented:
@mlaventure  @tonistiigi  sorry I am so busy these days and I have worked out another way to solve the Image Operation Competition. 
And I will submit the new PR today.

moypray commented:
@mlaventure @tonistiigi  I have updated this PR.

moypray committed:
Fix delete a image while saving it, delete successfully but failed to save it

Issue Description:
 * 1. Saving more than one images, `docker save -o a.tar aaa bbb`
 * 2. Delete the last image which in saving progress.  `docker rmi bbb`

Espected:
 Saving images operation shouldn't be disturbed. But the real result is that failed to
save image and get an error as below:
  `Error response from daemon: open
/var/lib/docker/image/devicemapper/imagedb/content/sha256/7c24e4d533a76e801662ad1b7e6e06bc1204f80110b5623e96ba2877c51479a1:
no such file or directory`

Analysis:
  1. While saving more than one images, it will get all the image info from reference/imagestore,
     and then using the `cached data` to save the images to a tar file.
  2. But this process doesn't have a resource lock, if a deletion operation comes, the image will be deleted,
     so saving operation will fail.

Solution:
 When begin to save an image, `Get` all the layers first. then the
deletion operation won't delete the layers.

Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

tonistiigi commented on image/tarexport/save.go:
```diff
@@ -23,14 +23,16 @@ import (
 type imageDescriptor struct {
 	refs   []reference.NamedTagged
 	layers []string
+	image  *image.Image
```
Keep the layer reference also in this object. Either leave both image and layers empty in `parseNamed` and fill them in a loop before `save` or if you fill them in `parseNamed` make sure they are released if `parseNamed` returns an error. Should avoid that one method returns image IDs that are already invalid for getting the layer refs by the time next method runs.

moypray commented:
Hi @tonistiigi  updated, thanks for your commets.

And I have done some basic tests:

1. prepare an images
```
root@localhost:~/bugs/save# cat Dockerfile
FROM ubuntu
MAINTAINER Wentao
ENV TEST  save_image
RUN echo helllo
```   
2. wrote a script:
```
root@localhost:~/bugs/save# cat a.sh
#!/bin/bash
docker save -o ../b.tar ubuntu  abc &
docker rmi abc
```
3. run `a.sh` to see the save operation status.

There are three kinds of status:
1. save image success, rmi success (rmi all the layers and image config)
2. save image success, rmi success (rmi only untag the image)
3. failed to save, rmi success.( rmi operation before caching the layers) (I think it is acceptable)
```
root@localhost:~/bugs/save# ./a.sh
Untagged: abc:latest
Deleted: sha256:e526a51d2b228c1dc2a8c327f528c221a88b9656967276740cb20e76ae175d1a
Deleted: sha256:f66a410427af6dd7480761c3a856f7b2883af4528da46bcfb8cb594f22a2509e
Deleted: sha256:79c2dca10a37972848dc31dc0812e719edac830f746afa091eed27772ac02596
Error response from daemon: No such image: abc
```

tonistiigi commented on image/tarexport/save.go:
```diff
@@ -93,33 +105,78 @@ func (l *tarexporter) parseNames(names []string) (map[image.ID]*imageDescriptor,
+	}
+
+	// Get all the layers for current image.
+	for i := range img.RootFS.DiffIDs {
```
You don't have to take a ref for every diffID, just `img.RootFS.ChainID`. The parent layers are protected is child ref is held(so you don't need a map as well).

tonistiigi commented on image/tarexport/save.go:
```diff
@@ -39,33 +41,45 @@ func (l *tarexporter) Save(names []string, outStream io.Writer) error {
 	imgDescr := make(map[image.ID]*imageDescriptor)
+	defer func() {
+		if rErr != nil {
+			l.releaseCachedLayers(imgDescr)
```
Instead of "cacheLayers", use "takeLayerReference", "releaseLayerReferences".

tonistiigi commented:
> There are three kinds of status:

If 1 & 2 have the same end state and only different output(after `wait %1`) then this is correct. Error on 3 is correct.

moypray commented:
@tonistiigi  update done.

tonistiigi commented on image/tarexport/save.go:
```diff
@@ -93,33 +105,68 @@ func (l *tarexporter) parseNames(names []string) (map[image.ID]*imageDescriptor,
+		return err
+	}
+	imgDescr.image = img
+	if imgDescr.cachedLayer != nil {
```
Why this check? Maybe it would be better if this returns `imageDescriptor` so it is clear that it can't be reused.

moypray replied:
What if the image is an empty image? I am not sure about the empty image. To be safe, here add a check.

tonistiigi replied:
What do you mean by "empty image"? afaics this function is always called with empty `imgDescr`. If you mean image with no layers then yes, it would be good to check that. But for that, you need to check for empty chainID.

moypray replied:
Sorry for late, We have 3 days holiday in China.

I thought if an image just have one empty layer, the `ChainID` may be emtpy string. But to check the code, it should be the empty layer ID. 

I think we can use `layerStore.Get` to Get an empty layer just like  [push](https://github.com/moby/moby/blob/master/distribution/push_v1.go#L217) operation.   So I think we do not need to check `nil` because of `EmptyLayer`. ( we can treat it as normal layer here)

But you mentioned that an image could have no layers and the ChainID should be empty.How does it happen? Could we use `imageStore.Get` to get this ? If couldn't, I think we should add a check here.

If I am wrong , correct me. Thanks.

moypray replied:
@tonistiigi  Updated,  And tested to save an empty image：
```
droot@localhost:~/bugs/save# docker save -o a.tar empty
Error response from daemon: empty export - not implemented
root@localhost:~/bugs/save#
```

If don't check whether `ChianID` is empty or not, we will get an error:
```
root@localhost:~/bugs/save# docker save -o a empty
Error response from daemon: layer does not exist
root@localhost:~/bugs/save#
```

tonistiigi commented on image/tarexport/save.go:
```diff
@@ -21,8 +21,10 @@ import (
+	refs        []reference.NamedTagged
+	layers      []string
+	image       *image.Image
+	cachedLayer layer.Layer
```
nit: `layerRef` or `rootFS`

tonistiigi commented on image/tarexport/save.go:
```diff
@@ -39,33 +41,45 @@ func (l *tarexporter) Save(names []string, outStream io.Writer) error {
 }
 
-func (l *tarexporter) parseNames(names []string) (map[image.ID]*imageDescriptor, error) {
+func (l *tarexporter) parseNames(names []string) (desc map[image.ID]*imageDescriptor, rErr error) {
```
Add a comment that descriptors need to be released by the caller.

tonistiigi reviewed (commented):
Design LGTM, just a couple of more nits.

moypray commented:
What's the error on Windows??

thaJeztah commented:
ping @tonistiigi @stevvooe PTAL

tonistiigi commented:
@moypray https://github.com/moby/moby/pull/32293#discussion_r118556092

tonistiigi commented on image/tarexport/save.go:
```diff
@@ -93,33 +107,73 @@ func (l *tarexporter) parseNames(names []string) (map[image.ID]*imageDescriptor,
+		return err
+	}
+	imgDescr.image = img
+	if imgDescr.layerRef != nil {
```
This still confuses me. `imgDescr.layerRef` can never be non-nil as all `imgDescr` passed to this function are uninitialized. As mentioned earlier, this would become much clearer if this function just returns `imgDescr` instead of taking one as a parameter.

moypray replied:
This is to avoid to take one imageID twice. if `Get` one imageID twice and add it to the map as one.  when `Release`，it will just release once..  So the imageID layers will  have dingling references. To be safe, I check it here.

But actually, the `parseNames` will never make this happen at present.

tonistiigi replied:
Yes, even if this is called multiple times for the same image, `layerRef` will always be nil and multiple references would be taken(that is fine).

moypray replied:
I think to be safe, we should keep this check...

tonistiigi replied:
Safe for what? Future code changes? In that case, make it explicit that there is always a fresh `imageDescr` returned from this function so nobody can use this function in a wrong way. You can't predict how code will be changed in the future. Atm, it just confuses the reader as it looks like a protection for something but not quite sure what exactly.

moypray replied:
> You can't predict how code will be changed in the future

Delete it then?  it will work fine for current logic.

moypray replied:
Updated

tonistiigi commented on image/tarexport/save.go:
```diff
@@ -93,33 +107,73 @@ func (l *tarexporter) parseNames(names []string) (map[image.ID]*imageDescriptor,
+
+	topLayerID := img.RootFS.ChainID()
+	if topLayerID == "" {
+		imgDescr.layerRef = layer.EmptyLayer
```
`layer.EmptyLayer` isn't correct here. It is a variable for detecting a layer that didn't have any actual changes. But in this case, you don't have any layers. Just skip setting `layerRef` here and check for nil in the `Release` function.

moypray replied:
Hmm. This is fine. I will rework it.

tonistiigi commented:
Thanks @moypray !

LGTM

moypray commented:
ping @thaJeztah @stevvooe

cpuguy83 reviewed (approved):
LGTM
